src/definitions/x509/trust_anchor.rs

```diff
@@ -0,0 +1,317 @@
+                Ok(root_certificate) => {
+                    errors.append(&mut process_validation_outcomes(
+                        leaf_certificate,
+                        root_certificate,
+                        rule_set,
+                    ));
+                }
+                Err(e) => {
+                    errors.push(e.into());
+                }
+            };
+        }
+        TrustAnchor::Custom(_certificate, _ruleset) => {
+            //TODO
+        }
+    }
+    errors
+}
+
+pub fn validate_with_trust_anchor(leaf_x509: X509, trust_anchor: TrustAnchor) -> Vec<X509Error> {
+    let mut errors: Vec<X509Error> = vec![];
+    let leaf_certificate = x509_cert::Certificate::from_der(&leaf_x509.bytes);
+
+    match leaf_certificate {
+        Ok(leaf) => {
+            errors.append(&mut validate_with_ruleset(leaf, trust_anchor));
+        }
+        Err(e) => errors.push(e.into()),
+    }
+    errors
+}
+
+pub fn check_validity_period(certificate: &CertificateInner) -> Vec<X509Error> {
+    let validity = certificate.tbs_certificate.validity;
+    let mut errors: Vec<X509Error> = vec![];
+    if validity.not_after.to_unix_duration().as_secs()
+        < OffsetDateTime::now_utc().unix_timestamp() as u64
+    {
+        errors.push(X509Error::ValidationError(format!(
+            "Expired certificate with subject: {:?}",
+            certificate.tbs_certificate.subject
+        )));
+    };
+    if validity.not_before.to_unix_duration().as_secs()
+        > OffsetDateTime::now_utc().unix_timestamp() as u64
+    {
+        errors.push(X509Error::ValidationError(format!(
+            "Not yet valid certificate with subject: {:?}",
+            certificate.tbs_certificate.subject
+        )));
+    };
+
+    errors
+}
+
+/* Validates:
+
+- all the correct distinghuished names are present
+and match the
+- all the correct extensions are present
+- the extensions are set to the ruleset values
+-  */
+fn apply_ruleset(
+    leaf_certificate: CertificateInner,
+    root_certificate: CertificateInner,
+    rule_set: ValidationRuleSet,
+) -> Result<Vec<X509Error>, X509Error> {
+    let mut errors: Vec<X509Error> = vec![];
+    // collect all the distinguished names in the root certificate that the validation ruleset requires
+    let root_distinguished_names: Vec<AttributeTypeAndValue> = root_certificate
+        .tbs_certificate
+        .subject
+        .0
+        .into_iter()
+        .map(|rdn| {
+            rdn.0
+                .into_vec()
+                .into_iter()
+                .filter(|atv| {
+                    rule_set
+                        .distinguished_names
+                        .iter()
+                        .any(|oid| oid == &atv.oid.to_string())
+                })
+                .collect::<Vec<AttributeTypeAndValue>>()
+        })
+        .collect::<Vec<Vec<AttributeTypeAndValue>>>()
+        .into_iter()
+        .flatten()
+        .collect();
```

Comment:
You should be able to do something like this which is much more efficient, and more readable in my opinion.
```suggestion
    let root_distinguished_names: Vec<Oid> = root_certificate
        .tbs_certificate
        .subject
        .0
        .into_iter()
        .map(|rdn| {
            rdn.0
                .into_iter()
                .map(|atv| {
                    atv.oid
                })
        })
        .flatten()
        .collect();
    // if all the needed distinguished names have been collected,
    // there should be the same number of names collected as are present in the ruleset
    if !rule_set.distinguished_names.iter().all(|oid| root_distinguished_names.contains(oid)) {
        errors.push(X509Error::ValidationError("The congifured validation ruleset requires a distinguished name that is not found in the submitted root certificate".to_string()));
    }
```